File: frame/map-frame-project/MapFrameCoreModule/src/main/java/com/grandtech/mapframe/core/editor/gc/BaseGeoSource.java
package com.grandtech.mapframe.core.editor.gc;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.grandtech.mapframe.core.editor.IGcCallBack;
import com.grandtech.mapframe.core.enumeration.GeometryType;
import com.grandtech.mapframe.core.util.CacheBoxUtil;
import com.grandtech.mapframe.core.util.FeatureUtil;
import com.grandtech.mapframe.core.util.Transformation;
import com.mapbox.geojson.Feature;
import com.mapbox.geojson.FeatureCollection;
import com.mapbox.mapboxsdk.style.expressions.Expression;
import com.mapbox.mapboxsdk.style.sources.GeoJsonSource;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.ObservableOnSubscribe;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * @ClassName BaseGeoSource
 * @Description TODO 临时图层
 * @Author: fs
 * @Date: 2021/4/20 9:16
 * @Version 2.0
 */

public  class BaseGeoSource extends GeoJsonSource implements IBaseGeoSource {

    private final String[] keyWords = {"uuid"};
    /**
     * 缓存GeoJson
     */
    private CacheBoxUtil cacheBoxUtil;
    /**
     * 缓存GeoJson路径
     */
    private String cachePath;

    private FeatureCollection featureCollection;

    public BaseGeoSource(String id) {
        super(id);
    }

    public BaseGeoSource(String id, String cachePath) {
        super(id);
        setSourceUrl(cachePath).initDraw();
    }

    public BaseGeoSource setSourceUrl(String cachePath) {
        this.cachePath = cachePath;
        return this;
    }


    public void initDraw() {
        if (cachePath != null) {
            cacheBoxUtil = CacheBoxUtil.get(new File(cachePath), Long.MAX_VALUE, Integer.MAX_VALUE);
        }
        featureCollection = FeatureCollection.fromJson(query());
        if (featureCollection == null) {
            featureCollection = FeatureCollection.fromFeatures(new ArrayList<Feature>());
        }
        this.setGeoJson(featureCollection);
    }

    @Override
    public String query() {
        String json = null;
        if (cacheBoxUtil != null) {
            json = queryCache();
        }
        return json;
    }


    @Override
    public String queryCache() {
        return cacheBoxUtil.getAsString(getId());
    }

    @Override
    public Feature queryFeature(String key, String value) {
        return FeatureUtil.queryFeatureFormCollection(featureCollection, key, value);
    }

    @NonNull
    @Override
    public List<Feature> querySourceFeatures(@Nullable Expression filter) {
        return super.querySourceFeatures(filter);
    }

    @NonNull
    public Feature querySourceFeature(String key, String val) {
        Feature feature = FeatureUtil.queryFeatureFormCollection(featureCollection, key, val);
        return feature;
    }

    @NonNull
    public List<Feature> queryAllSourceFeatures() {

        return featureCollection.features();
    }

    @Override
    public Feature insert(Feature feature) {
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            feature = this.insertCache(feature);
        }
        return feature;
    }


    @Override
    public Feature insertCache(Feature feature) {
        featureCollection.features().add(feature);
        String json = Transformation.mapBoxObj2Json(featureCollection);
        cacheBoxUtil.put(getId(), json);
        this.setGeoJson(featureCollection);
        return feature;
    }

    @Override
    public FeatureCollection inserts(List<Feature> features) {
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            final String id =getId();
            featureCollection.features().addAll(features);
            String json = Transformation.mapBoxObj2Json(featureCollection);
            cacheBoxUtil.put(id, json);
        }
        return featureCollection;
    }
    @Override
    public void insertsCacheSync(final List<Feature> features, IGcCallBack iGcCallBack) {
        String id =getId();
        Observable.create((ObservableOnSubscribe<FeatureCollection>) emitter -> {
            featureCollection.features().addAll(features);
            String json = Transformation.mapBoxObj2Json(featureCollection);
            cacheBoxUtil.put(id, json);
            emitter.onNext(featureCollection);
            emitter.onComplete();
        }).subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(new Observer<FeatureCollection>() {
                    @Override
                    public void onSubscribe(@io.reactivex.annotations.NonNull Disposable d) {

                    }

                    @Override
                    public void onNext(@io.reactivex.annotations.NonNull FeatureCollection featureCollection) {
                        setGeoJson(featureCollection);
                        iGcCallBack.onGcCallBack(features);
                    }

                    @Override
                    public void onError(@io.reactivex.annotations.NonNull Throwable e) {
                        e.fillInStackTrace();
                        iGcCallBack.onError( e);
                    }

                    @Override
                    public void onComplete() {

                    }
                });
    }

    @Override
    public Boolean delete(Feature feature) {
        boolean flag = false;
        if (cacheBoxUtil == null) {
            return flag;
        }
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            flag = this.deleteCache(feature);
        }
        return flag;
    }

    @Override
    public Boolean deleteCache(Feature feature) {
        boolean flag = false;
        int delIndex = FeatureUtil.queryFeatureFormCollection(featureCollection, feature, keyWords);
        if (delIndex != -1) {
            featureCollection.features().remove(delIndex);
            this.setGeoJson(featureCollection);
            //更新缓存数据
            String json = Transformation.mapBoxObj2Json(featureCollection);
            cacheBoxUtil.put(getId(), json);
            flag = true;
        }
        return flag;
    }

    @Override
    public Boolean deleteFeatures(@NonNull Map<String, Object> con) {
        boolean flag = false;
        if (cacheBoxUtil == null) {
            return flag;
        }
        List<Feature> features = queryFeatures(con);
        for(Feature feature : features){
            flag = deleteCache(feature);
        }
        return flag;
    }

    @Override
    public Boolean deleteAll() {
        boolean res = false;
        if ( cacheBoxUtil == null) {
            return null;
        }
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            featureCollection.features().clear();
            this.setGeoJson(featureCollection);
            String json = Transformation.mapBoxObj2Json(featureCollection);
            cacheBoxUtil.put(getId(), json);
            res = true;
        }
        return res;
    }

    @Override
    public Feature update(Feature feature) {
        if (cacheBoxUtil == null) {
            return null;
        }
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            feature = updateCache(feature);
        }
        return feature;
    }

    @Override
    public Feature updateCache(Feature feature) {
        int index = FeatureUtil.queryFeatureFormCollection(featureCollection, feature);
        if (index != -1) {
            featureCollection.features().set(index, feature);
            this.setGeoJson(featureCollection);
            //更新缓存数据
            String json = Transformation.mapBoxObj2Json(featureCollection);
            cacheBoxUtil.put(getId(), json);
            return feature;
        } else {
            return null;
        }
    }

    @Override
    public List<Feature> queryFeatures(Map<String, Object> con) {
        return FeatureUtil.queryFeatureCollection(featureCollection, con);
    }

    ;
    @Override
    public List<Feature> queryFeaturesLike(Map<String, Object> con) {
        return FeatureUtil.queryFeatureCollectionLike(featureCollection, con);
    }
}

File: frame/map-frame-project/MapFrameCoreModuleV10/src/main/java/com/gradtech/mapframev10/core/offline/geosource/BaseGeoSource.kt
package com.gradtech.mapframev10.core.offline.geosource

import android.util.Log
import androidx.lifecycle.LifecycleCoroutineScope
import com.gradtech.mapframev10.core.util.CacheBoxUtil
import com.gradtech.mapframev10.core.util.FeatureUtil
import com.gradtech.mapframev10.core.util.Transformation
import com.mapbox.geojson.Feature
import com.mapbox.geojson.FeatureCollection
import com.mapbox.maps.extension.style.sources.generated.GeoJsonSource
import com.mapbox.maps.extension.style.sources.generated.geoJsonSource
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jetbrains.annotations.NotNull
import java.io.File
import java.util.*

/**
 * @ClassName BaseGeoSource
 * @Description TODO
 * @Author: fs
 * @Date: 2022/9/16 9:57
 * @Version 2.0
 */
class BaseGeoSource : IBaseGeoSource {

    private val keyWords = arrayOf("uuid")

    /**
     * 缓存GeoJson
     */
    private var cacheBoxUtil: CacheBoxUtil? = null

    private var id: String

    /**
     * 缓存GeoJson路径
     */
    private var cachePath: String

    private var featureCollection: FeatureCollection? = null

    private var geojosnSource: GeoJsonSource? = null

    @JvmOverloads
    constructor(@NotNull id: String, @NotNull cachePath: String) {
        this.id = id
        this.cachePath = cachePath
        initDraw()
    }

    override fun setSourceUrl(cachePath: String): IBaseGeoSource {
        this.cachePath = cachePath
        return this
    }

    override fun getGeoSource(): GeoJsonSource? = geojosnSource;

    override fun initDraw() {
        if (geojosnSource == null) {
            geojosnSource = geoJsonSource(id)
        }
        if (cachePath != null) {
            cacheBoxUtil = CacheBoxUtil.get(File(cachePath), Long.MAX_VALUE, Int.MAX_VALUE)
        }
        val query = query()
        if (query != null) {
            featureCollection = FeatureCollection.fromJson(query()!!)
        }
        if (featureCollection == null) {
            featureCollection = FeatureCollection.fromFeatures(ArrayList())
        }
        if(featureCollection!!.features()!!.size>0){
            geojosnSource?.featureCollection(featureCollection!!)
        }

    }

    override fun query(): String? {
        var json: String? = null
        if (cacheBoxUtil != null) {
            json = queryCache()
        }
        return json
    }

    override fun queryCache(): String? {
        return cacheBoxUtil!!.getAsString(geojosnSource?.sourceId)
    }

    override fun queryFeatures(con: HashMap<String, Any>): List<Feature> {
        return FeatureUtil.queryFeatureCollection(featureCollection, con)
    }

    override fun queryFeaturesLike(con: Map<String, Any>): List<Feature> {
        return FeatureUtil.queryFeatureCollectionLike(featureCollection, con)
    }

    override fun queryFeature(key: String, value: String): Feature? {
        return FeatureUtil.queryFeatureFormCollection(featureCollection, key, value)
    }

    override fun insert(feature: Feature): Feature? {
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            val feature = insertCache(feature)
            return feature
        }
        return null
    }

    override fun insertCache(feature: Feature): Feature {
        featureCollection!!.features()!!.add(feature)
        val json: String = Transformation.mapBoxObj2Json(featureCollection)
        cacheBoxUtil!!.put(id, json)
        geojosnSource?.featureCollection(featureCollection!!)
        return feature
    }

    override fun inserts(features: List<Feature>): FeatureCollection? {
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            featureCollection!!.features()!!.addAll(features)
            val json: String = Transformation.mapBoxObj2Json(featureCollection)
            cacheBoxUtil!!.put(id, json)
            geojosnSource?.featureCollection(featureCollection!!)
        }
        return featureCollection
    }

    override fun insertsCacheUnSync(features: List<Feature>, iGcCallBack: IBaseGeoSource.IGcCallBack?) {
        // 协程构建器         主线程
        val launch = GlobalScope.launch(Dispatchers.Main) {
            //任务调度器    子线程
            val featureCollection = withContext(Dispatchers.IO) {
                inserts(features)
            }
            if (featureCollection != null) {
                iGcCallBack?.onGcCallBack(featureCollection.features())
            } else {
                iGcCallBack?.onError(Throwable())
            }

        }
      //  launch.start()

    }

    override fun delete(feature: Feature): Boolean {
        var flag = false
        if (cacheBoxUtil == null) {
            return flag
        }
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            flag = deleteCache(feature)
        }
        return flag
    }

    override fun deleteCache(feature: Feature): Boolean {
        var flag = false
        val delIndex: Int = FeatureUtil.queryFeatureFormCollection(featureCollection, feature, *keyWords)
        if (delIndex != -1) {
            featureCollection!!.features()!!.removeAt(delIndex)
            this.geojosnSource?.featureCollection(featureCollection!!)
            //更新缓存数据
            val json: String = Transformation.mapBoxObj2Json(featureCollection)
            cacheBoxUtil!!.put(id, json)
            flag = true
        }
        return flag
    }

    override fun deleteFeatures(con: HashMap<String, Any>): Boolean {
        var flag = false
        if (cacheBoxUtil == null) {
            return flag
        }
        val features = queryFeatures(con)
        for (feature in features) {
            flag = deleteCache(feature)
        }
        return flag
    }

    override fun deleteAll(): Boolean {
        var res = false
        if (cacheBoxUtil == null) {
            return false
        }
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            featureCollection!!.features()!!.clear()
            this.geojosnSource?.featureCollection(featureCollection!!)
            val json: String = Transformation.mapBoxObj2Json(featureCollection)
            cacheBoxUtil!!.put(id, json)
            res = true
        }
        return res
    }

    override fun update(feature: Feature?): Boolean {
        if (cacheBoxUtil == null) {
            return false
        }
        if (cacheBoxUtil != null) {
            //更新内存和缓存
            updateCache(feature)
            return true;
        }
        return false
    }

    override fun updateCache(feature: Feature?): Boolean {
        val index: Int = FeatureUtil.queryFeatureFormCollection(featureCollection, feature)
        return if (index != -1) {
            featureCollection!!.features()!![index] = feature
            this.geojosnSource?.featureCollection(featureCollection!!)
            //更新缓存数据
            val json: String = Transformation.mapBoxObj2Json(featureCollection)
            cacheBoxUtil!!.put(id, json)
            true
        } else {
            false
        }
    }


}